Unable to Configure and Run camera_ros After Reinstallation

# Unable to Configure and Run `camera_ros` After System Reset

## Description
I previously used the `camera_ros` package successfully on my Raspberry Pi 5 without any additional configuration. However, after resetting my computer, I’ve been unable to get it working despite following the same steps and reinstalling all required dependencies.

The node fails to configure the camera and consistently throws errors. I also noticed that both `libcamera` and `camera_ros` have been updated recently, and reverting to older versions did not solve the problem. Additionally, I can still capture images using **QCam**, so `libcamera` appears to be functioning correctly.

---

## System Details
- **Hardware:** Raspberry Pi 5 + IMX708 camera module (2 units)  
- **Software:**
  - **OS:** Ubuntu 24.04  
  - **ROS 2:** Jazzy  
  - **Libraries:** `libcamera`, `v4l2`, `camera_ros`  

---

## Current Issues and Errors
1. **v4l2_camera Error:**  
 ```bash
Failed stream start: Invalid argument (22)
 ```
3. **Camera_ros Error:**  

 ```
Can't configure camera with invalid configuration
terminate called after throwing an instance of 'std::runtime_error'
what():  failed to configure streams
Abort.
 ```
## Current Issues and Errors
- **Error Summary:** The `camera_ros` node fails to configure the camera streams due to invalid configuration and unsupported pixel formats.
- **Details:**
  - The default pixel format (`YUV420`) is not supported by the camera.
  - The system attempts to auto-select a pixel format (`R8`) and resolution (`640x480`), but adjusts it to an unsupported configuration (`1536x864-SBGGR16`).
  - The stream configuration results in the error:  
    ```
    Can't configure camera with invalid configuration
    failed to configure streams
    Removing camera while still in use
    ```
  - Additionally, the container throws the following exception:  
    ```
    Component constructor threw an exception: failed to configure streams
    ```
## ros2 launch camera_ros  camera.launch.py Error
- **Error Summary:** The `camera_ros` node fails to configure the camera due to invalid stream configuration and unsupported default pixel formats.
- **Details:**
  - The default pixel format (`YUV420`) is not supported by the camera.
  - The system attempts to auto-select a pixel format (`R8`) and resolution (`1920x1080`), but adjusts it to an unsupported configuration (`2304x1296-SBGGR16`).
  - This leads to the following critical error:
    ```
    Can't configure camera with invalid configuration
    failed to configure streams
    ```
  - The process terminates with:
    ```
    terminate called after throwing an instance of 'std::runtime_error'
    what():  failed to configure streams
    ```
## ros2 run camera_ros camera_node   Errors
- **Error Summary:** The `camera_ros` node fails to configure the camera due to invalid stream configuration and unsupported default pixel formats.
- **Details:**
  - The default pixel format (`YUV420`) is not supported by the camera.
  - The system attempts to auto-select a pixel format (`R8`) and resolution (`1920x1080`), but adjusts it to an unsupported configuration (`2304x1296-SBGGR16`).
  - This leads to the following critical error:
    ```
    Can't configure camera with invalid configuration
    failed to configure streams
    ```
  - The process terminates with:
    ```
    terminate called after throwing an instance of 'std::runtime_error'
    what():  failed to configure streams
    ```


## Steps to Reproduce
1. Installed all dependencies for `camera_ros` and `libcamera`.
2. Verified the camera works with QCam.
4. Attempted to run `camera_ros` using the following commands:
 ```bash
 ros2 run camera_ros camera_node --ros-args -p camera:=0 -p format:="YUYV" -p width:=1640 -p height:=1232 
 ```
or: 
 ```bash
ros2 run camera_ros camera_node --ros-args -p camera:=0 
```


5. Tried modifying the role parameter to raw and viewfinder, but the errors persisted.
6. Reverted both libcamera and camera_ros to previous versions.


## What I Tried
- Verified that QCam can still capture images, confirming that `libcamera` works.
- Used supported formats and resolutions (e.g., `1640x1232-YUYV`).
- Reinstalled all dependencies.
- Tried previous versions of both `libcamera` and `camera_ros`.

## Expected Behavior
The node should configure the camera successfully and stream images, as it did during the initial setup without any manual adjustments.

## Questions
1. Are there any known issues with the latest updates to `libcamera` or `camera_ros`?
2. How can I prevent these configuration errors?
3. Is there a recommended debugging or setup process to ensure compatibility between `libcamera` and `camera_ros`?

## Additional Context
The current situation has become very frustrating, as the same hardware and software setup worked flawlessly during my initial installation. Any help or guidance would be greatly appreciated.

#### **Hardware and Software Used**

> Device: Raspberry Pi 5
> Camera: Raspberry Pi Camera Module 3
> Operating Systems: `Ubuntu 24.04` (64-bit) arm
> Libcamera Version: v0.3.0
> ROS 2 Version:` Jazzy`

#### **Findings**
Camera Not Detected on Ubuntu:
The `"Available cameras"` section is empty.
The error `"Unable to acquire a Unicam instance"` was encountered.
Camera Works Seamlessly on Raspbian:
The camera is detected, and image capture is successful.
The ROS 2 node runs without issues.

Driver Compatibility:
It was observed that camera drivers (especially Unicam) are not functioning correctly on Ubuntu.

The Raspberry Pi Camera Module 3 is not properly detected on Ubuntu, whereas it works seamlessly on Raspbian. This indicates a potential incompatibility with camera drivers or the libcamera configuration on Ubuntu.

LIBCANERA_LOG_LEVELS=*:DEBUG cam -l
```text
[1:11:25.540318062] [5901] DEBUG IPAModule ipa_module.cpp:334 ipa_rpi_vc4.so: IPA module /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so is signed
[1:11:25.540429339] [5901] DEBUG IPAManager ipa_manager.cpp:245 Loaded IPA module '/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so'
[1:11:25.540470987] [5901]  INFO Camera camera_manager.cpp:313 libcamera v0.3.0
[1:11:25.540642060] [5904] DEBUG Camera camera_manager.cpp:69 Starting camera manager
[1:11:25.568831078] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media0
[1:11:25.570020219] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media0 due to 10 missing dependencies
[1:11:25.573099571] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media0 found
[1:11:25.573152033] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media0: rp1-cfe
[1:11:25.573346588] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media1
[1:11:25.573655327] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media1 due to 10 missing dependencies
[1:11:25.575633537] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media1 found
[1:11:25.575648370] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media1: rp1-cfe
[1:11:25.575769092] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rpivid" created from /dev/media3
[1:11:25.575779444] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media3 due to 1 missing dependencies
[1:11:25.576069646] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media3 found
[1:11:25.576092034] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media3: rpivid
[1:11:25.576450699] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media2
[1:11:25.576489069] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media2 due to 9 missing dependencies
[1:11:25.576858363] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media4
[1:11:25.576922789] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media4 due to 9 missing dependencies
[1:11:25.579778716] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media2 found
[1:11:25.579868216] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media2: pispbe
[1:11:25.583751767] [5904] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media4 found
[1:11:25.583789174] [5904] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media4: pispbe
[1:11:25.584485355] [5904] DEBUG Camera camera_manager.cpp:134 Found registered pipeline handler 'rpi/vc4'
[1:11:25.584546762] [5904] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
[1:11:25.584554836] [5904] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
Available cameras:

```

~/ros2_ws$ ros2 run camera_ros camera_node 

```text
[1:07:20.248118571] [5694] DEBUG IPAModule ipa_module.cpp:334 ipa_rpi_vc4.so: IPA module /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so is signed
[1:07:20.248228923] [5694] DEBUG IPAManager ipa_manager.cpp:245 Loaded IPA module '/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so'
[1:07:20.279787433] [5694]  INFO Camera camera_manager.cpp:313 libcamera v0.3.0
[1:07:20.280057192] [5709] DEBUG Camera camera_manager.cpp:69 Starting camera manager
[1:07:20.306533728] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media0
[1:07:20.306835877] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media0 due to 10 missing dependencies
[1:07:20.308694918] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media0 found
[1:07:20.308708622] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media0: rp1-cfe
[1:07:20.308849844] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media1
[1:07:20.309106900] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media1 due to 10 missing dependencies
[1:07:20.310955589] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media1 found
[1:07:20.310965126] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media1: rp1-cfe
[1:07:20.311068849] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rpivid" created from /dev/media3
[1:07:20.311078034] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media3 due to 1 missing dependencies
[1:07:20.311264034] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media3 found
[1:07:20.311272164] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media3: rpivid
[1:07:20.311386016] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media2
[1:07:20.311395905] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media2 due to 9 missing dependencies
[1:07:20.311512165] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media4
[1:07:20.311525128] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media4 due to 9 missing dependencies
[1:07:20.313136353] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media2 found
[1:07:20.313146279] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media2: pispbe
[1:07:20.314749134] [5709] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media4 found
[1:07:20.314760634] [5709] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media4: pispbe
[1:07:20.315274969] [5709] DEBUG Camera camera_manager.cpp:134 Found registered pipeline handler 'rpi/vc4'
[1:07:20.315333432] [5709] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
[1:07:20.315341821] [5709] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
terminate called after throwing an instance of 'std::runtime_error'
  what():  no cameras available
[ros2run]: Aborted
```

~/ros2_ws$ ros2 launch camera_ros  camera.launch.py 

``` text
[INFO] [launch]: All log files can be found below /home/pixbot/.ros/log/2024-11-30-10-40-33-695463-pixbot-desktop-5740
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [component_container-1]: process started with pid [5758]
[component_container-1] [INFO] [1732952434.412683348] [camera_container]: Load Library: /home/pixbot/ros2_ws/install/camera_ros/lib/libcamera_component.so
[component_container-1] [INFO] [1732952434.526415786] [camera_container]: Found class: rclcpp_components::NodeFactoryTemplate<camera::CameraNode>
[component_container-1] [INFO] [1732952434.526530490] [camera_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<camera::CameraNode>
[component_container-1] [1:07:52.938455688] [5758] DEBUG IPAModule ipa_module.cpp:334 ipa_rpi_vc4.so: IPA module /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so is signed
[component_container-1] [1:07:52.938656466] [5758] DEBUG IPAManager ipa_manager.cpp:245 Loaded IPA module '/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so'
[component_container-1] [1:07:52.957544709] [5758]  INFO Camera camera_manager.cpp:313 libcamera v0.3.0
[component_container-1] [1:07:52.957843154] [5774] DEBUG Camera camera_manager.cpp:69 Starting camera manager
[component_container-1] [1:07:52.982661686] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media0
[component_container-1] [1:07:52.984111875] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media0 due to 10 missing dependencies
[component_container-1] [1:07:52.986655380] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media0 found
[component_container-1] [1:07:52.986707713] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media0: rp1-cfe
[component_container-1] [1:07:52.986954899] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rp1-cfe" created from /dev/media1
[component_container-1] [1:07:52.987611456] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media1 due to 10 missing dependencies
[component_container-1] [1:07:52.990527925] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media1 found
[component_container-1] [1:07:52.990592999] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media1: rp1-cfe
[component_container-1] [1:07:52.990921518] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "rpivid" created from /dev/media3
[component_container-1] [1:07:52.990964148] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media3 due to 1 missing dependencies
[component_container-1] [1:07:52.991720261] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media3 found
[component_container-1] [1:07:52.991784298] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media3: rpivid
[component_container-1] [1:07:52.992453633] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media2
[component_container-1] [1:07:52.992522373] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media2 due to 9 missing dependencies
[component_container-1] [1:07:52.993176282] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:230 New media device "pispbe" created from /dev/media4
[component_container-1] [1:07:52.993277245] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:96 Defer media device /dev/media4 due to 9 missing dependencies
[component_container-1] [1:07:52.997043531] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media2 found
[component_container-1] [1:07:52.997086938] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media2: pispbe
[component_container-1] [1:07:52.998927109] [5774] DEBUG DeviceEnumerator device_enumerator_udev.cpp:322 All dependencies for media device /dev/media4 found
[component_container-1] [1:07:52.998950257] [5774] DEBUG DeviceEnumerator device_enumerator.cpp:258 Added device /dev/media4: pispbe
[component_container-1] [1:07:52.999594962] [5774] DEBUG Camera camera_manager.cpp:134 Found registered pipeline handler 'rpi/vc4'
[component_container-1] [1:07:52.999684203] [5774] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
[component_container-1] [1:07:52.999691110] [5774] DEBUG RPI vc4.cpp:178 Unable to acquire a Unicam instance
[component_container-1] [ERROR] [1732952434.625048527] [camera_container]: Component constructor threw an exception: no cameras available
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node '' of type 'camera::CameraNode' in container '/camera_container': Component constructor threw an exception: no cameras available
[component_container-1] [INFO] [1732952434.632353912] [camera_container]: Load Library: /opt/ros/jazzy/lib/libimage_view_nodes.so
[component_container-1] [INFO] [1732952434.745858572] [camera_container]: Found class: rclcpp_components::NodeFactoryTemplate<image_view::DisparityViewNode>
[component_container-1] [INFO] [1732952434.745932924] [camera_container]: Found class: rclcpp_components::NodeFactoryTemplate<image_view::ExtractImagesNode>
[component_container-1] [INFO] [1732952434.745946517] [camera_container]: Found class: rclcpp_components::NodeFactoryTemplate<image_view::ImageSaverNode>
[component_container-1] [INFO] [1732952434.745959906] [camera_container]: Found class: rclcpp_components::NodeFactoryTemplate<image_view::ImageViewNode>
[component_container-1] [INFO] [1732952434.745972350] [camera_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<image_view::ImageViewNode>
[component_container-1] [INFO] [1732952434.765416372] [image_view_node]: Using transport "raw"
[INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/image_view_node' in container '/camera_container'
[component_container-1] QSocketNotifier: Can only be used with threads started with QThread
```

ls /dev/video*
```text
/dev/video0   /dev/video14  /dev/video22  /dev/video28  /dev/video33  /dev/video5
/dev/video1   /dev/video15  /dev/video23  /dev/video29  /dev/video34  /dev/video6
/dev/video10  /dev/video19  /dev/video24  /dev/video3   /dev/video35  /dev/video7
/dev/video11  /dev/video2   /dev/video25  /dev/video30  /dev/video36  /dev/video8
/dev/video12  /dev/video20  /dev/video26  /dev/video31  /dev/video37  /dev/video9
/dev/video13  /dev/video21  /dev/video27  /dev/video32  /dev/video4
```


v4l2-ctl --list-devices

```text
pispbe (platform:1000880000.pisp_be):
    /dev/video20
    /dev/video21
    /dev/video22
    /dev/video23
    /dev/video24
    /dev/video25
    /dev/video26
    /dev/video27
    /dev/video28
    /dev/video29
    /dev/video30
    /dev/video31
    /dev/video32
    /dev/video33
    /dev/video34
    /dev/video35
    /dev/video36
    /dev/video37
    /dev/media2
    /dev/media4
rp1-cfe (platform:1f00110000.csi):
    /dev/video8
    /dev/video9
    /dev/video10
    /dev/video11
    /dev/video12
    /dev/video13
    /dev/video14
    /dev/video15
    /dev/media0
rp1-cfe (platform:1f00128000.csi):
    /dev/video0
    /dev/video1
    /dev/video2
    /dev/video3
    /dev/video4
    /dev/video5
    /dev/video6
    /dev/video7
    /dev/media1
rpivid (platform:rpivid):
    /dev/video19
    /dev/media3
```

--- a/README.md
+++ b/README.md
@@ -7,13 +7,16 @@ This ROS 2 node provides support for a variety of cameras via [libcamera](https:
 
 Binary packages are available via the ROS package repository for some Linux and ROS distributions (check with `rosdep resolve camera_ros`). If it's available, you can install the DEB or RPM packages via:
 ```sh
-# source ROS distribution
-source /opt/ros/humble/setup.bash
+# source a ROS distribution
+source /opt/ros/$ROS_DISTRO/setup.bash
 # DEB package
 sudo apt install ros-$ROS_DISTRO-camera-ros
 # RPM package
 sudo dnf install ros-$ROS_DISTRO-camera-ros
 ```
+
+> [!NOTE]
+> This also installs the package [`libcamera`](https://index.ros.org/r/libcamera/) as dependency. This is the bloomed version of the official upstream repo at https://git.libcamera.org/libcamera/libcamera.git and may not contain full support for all Raspberry Pi camera modules. If you need full camera module support on Raspberry Pi, you have to build the "raspberrypi" fork from https://github.com/raspberrypi/libcamera manually.
 
 
 ## Build Instructions

--- a/README.md
+++ b/README.md
@@ -2,9 +2,8 @@
 
 This ROS 2 node provides support for a variety of cameras via [libcamera](https://libcamera.org). Amongst others, this node supports V4L2 and [Raspberry Pi cameras](https://www.raspberrypi.com/documentation/computers/camera_software.html).
 
-## Install
 
-### Binary
+## Installation
 
 Binary packages are available via the ROS package repository for some Linux and ROS distributions (check with `rosdep resolve camera_ros`). If it's available, you can install the DEB or RPM packages via:
 ```sh
@@ -16,9 +15,10 @@ sudo apt install ros-$ROS_DISTRO-camera-ros
 sudo dnf install ros-$ROS_DISTRO-camera-ros
 ```
 
-### Source
 
-#### libcamera dependency
+## Build Instructions
+
+### libcamera dependency
 
 The `camera_ros` node depends on libcamera version 0.1 or later.
 
@@ -32,7 +32,7 @@ sudo apt install pkg-config python3-yaml python3-ply python3-jinja2 openssl liby
 sudo dnf install pkgconfig python3-yaml python3-ply python3-jinja2 openssl libyaml-devel openssl-devel libudev-devel libatomic meson
 ```
 
-#### build camera_ros
+### build camera_ros
 
 The `camera_ros` package is built in a colcon workspace. The following instructions assume that you are building libcamera from source in the colcon workspace.
 
@@ -54,6 +54,7 @@ colcon build
 
 If you installed libcamera externally, you can omit the `colcon-meson` and `libcamera` steps. Additionally, if there is a binary package and a rosdep entry for libcamera (check with `rosdep resolve libcamera`) you can also omit `--skip-keys=libcamera` and have this binary dependency resolved automatically.
 
+
 ## Launching the Node
 
 The package provides a standalone node executable:
@@ -68,6 +69,7 @@ and an example launch file for the composable node:
 ```sh
 ros2 launch camera_ros camera.launch.py
 ```
+
 
 ## Interfaces
 

--- a/README.md
+++ b/README.md
@@ -34,27 +34,32 @@ The `camera_ros` node depends on libcamera version 0.1 or later. There are diffe
     2. Install `colcon-meson` via the package manager, `sudo apt install -y python3-colcon-meson`, or pip, `pip install colcon-meson`.
 
 
-### build camera_ros
+### camera_ros
 
-The `camera_ros` package is built in a colcon workspace. The following instructions assume that you are building libcamera from source in the colcon workspace.
-
+The `camera_ros` package is built together with libcamera in a colcon workspace:
 ```sh
-# create workspace with camera_ros package
-mkdir -p ~/camera_ws/
-cd ~/camera_ws/
-git clone https://github.com/christianrauch/camera_ros.git src/camera_ros
+# create workspace
+mkdir -p ~/camera_ws/src
+cd ~/camera_ws/src
 
-# optional: build libcamera in colcon workspace
-pip install colcon-meson
-git clone https://git.libcamera.org/libcamera/libcamera.git src/libcamera
+# check out libcamera
+# Option A: official upstream
+git clone https://git.libcamera.org/libcamera/libcamera.git
+# Option B: raspberrypi fork with support for newer camera modules
+# git clone https://github.com/raspberrypi/libcamera.git
+
+# check out this camera_ros repository
+git clone https://github.com/christianrauch/camera_ros.git
 
 # resolve binary dependencies and build workspace
-source /opt/ros/humble/setup.bash
+source /opt/ros/$ROS_DISTRO/setup.bash
+cd ~/camera_ws/
 rosdep install --from-paths src --ignore-src --skip-keys=libcamera
 colcon build
 ```
 
-If you installed libcamera externally, you can omit the `colcon-meson` and `libcamera` steps. Additionally, if there is a binary package and a rosdep entry for libcamera (check with `rosdep resolve libcamera`) you can also omit `--skip-keys=libcamera` and have this binary dependency resolved automatically.
+If you are using a binary distribution of libcamera, you can skip adding this to the workspace. Additionally, if you want to use the bloomed libcamera package in the ROS repos, you can also omit `--skip-keys=libcamera` and have this binary dependency resolved automatically.
+
 
 
 ## Launching the Node

--- a/README.md
+++ b/README.md
@@ -21,19 +21,18 @@ sudo dnf install ros-$ROS_DISTRO-camera-ros
 
 ## Build Instructions
 
-### libcamera dependency
+### libcamera
 
-The `camera_ros` node depends on libcamera version 0.1 or later.
+The `camera_ros` node depends on libcamera version 0.1 or later. There are different ways to install this dependency:
 
-Some Linux and ROS distributions provide binary libcamera packages. Check your package manager for `libcamera` and `rosdep resolve libcamera` to see if binary packages are available.
+- __System Package:__ Most Linux distributions provide a binary libcamera package. However, the Raspberry Pi OS uses a [custom libcamera fork](https://github.com/raspberrypi/libcamera) with additional support for newer camera modules. When using the distribution packages, you have to skip the `libcamera` rosdep key when resolving dependencies (`rosdep install [...] --skip-keys=libcamera`).
 
-If your distribution does not provide a binary libcamera package, you have to compile libcamera from source either independent of the colcon workspace according to the [official build instructions](https://libcamera.org/getting-started.html) or as part of the colcon workspace. Either way, you have to install libcamera's build dependencies manually:
-```sh
-# DEB
-sudo apt install pkg-config python3-yaml python3-ply python3-jinja2 openssl libyaml-dev libssl-dev libudev-dev libatomic1 meson
-# RPM
-sudo dnf install pkgconfig python3-yaml python3-ply python3-jinja2 openssl libyaml-devel openssl-devel libudev-devel libatomic meson
-```
+- __ROS Package:__ You can also install a newer version from the ROS repo (package `ros-$ROS_DISTRO-libcamera`). This package will be installed by default when building `camera_ros` from source and resolving the rosdep keys.
+
+- __Source:__ Finally, you can always build libcamera from source. This is currently the only option for using the "raspberrypi" fork on Ubuntu. You can build libcamera as part of the ROS workspace using `colcon-meson`. This is recommended over a system-wide installation as it avoids conflicts with the system package. You will need to install the following dependencies:
+    1. Install the libcamera build dependencies according to https://libcamera.org/getting-started.html#dependencies.
+    2. Install `colcon-meson` via the package manager, `sudo apt install -y python3-colcon-meson`, or pip, `pip install colcon-meson`.
+
 
 ### build camera_ros
 

--- a/README.md
+++ b/README.md
@@ -168,3 +168,31 @@ and then run the node with libcamera and ROS debug information in `gdb`:
 ```sh
 LIBCAMERA_LOG_LEVELS=*:DEBUG ros2 run --prefix 'gdb -ex run --args' camera_ros camera_node --ros-args --log-level debug -p width:=640 -p height:=480
 ```
+
+
+### Common Issues
+
+#### No camera is detected (exception `no cameras available`)
+
+For standard V4L2 camera devices, check that its connection is detected (`lsusb`) and that it is also detected by V4L2 (`v4l2-ctl --list-devices`).
+
+On the Raspberry Pi, use `sudo vclog --msg` to inspect the VideoCore log messages for detected cameras. `vclog` is part of Raspberry Pi's `utils` repo at https://github.com/raspberrypi/utils. If this is not available in your distribution, such as Ubuntu, you have to build it from source:
+```sh
+# install build dependencies
+sudo apt -y install --no-install-recommends wget ca-certificates gcc libc6-dev
+# download vclog and build
+wget https://raw.githubusercontent.com/raspberrypi/utils/refs/heads/master/vclog/vclog.c
+gcc vclog.c -o vclog
+# show VideoCore log messages
+sudo ./vclog --msg
+```
+
+In the VideoCore log, you should find something like `Found camera 'ov5647' []` or `Found camera 'imx708' [...]`, which indicates the connected sensor, in these cases the _OmniVision OV5647_ and _Sony IMX708_. Look up this sensor in the [Raspberry Pi camera modules _Hardware Specification_](https://www.raspberrypi.com/documentation/accessories/camera.html#hardware-specification) and match the detected sensor with with the camera module. In this case, the _OmniVision OV5647_ image sensor belongs to the Camera Module 1 and the _Sony IMX708_ to the Camera Module 3.
+
+If you are using a Raspberry Pi Camera Module, make sure that it is supported by the libcamera variant that you have installed. For example, the official upstream libcamera supports the Camera Module 1 (OmniVision OV5647), while the Camera Module 3 (Sony IMX708) is currently only supported by the "raspberrypi" fork of libcamera. To verify that your camera is working with libcamera, build the `cam` example from the libcamera repo and list the camera devices with `cam -l`. For a camera with the _OmniVision OV5647_ sensor, this will show something like `'ov5647' (/base/axi/pcie@120000/rp1/i2c@88000/ov5647@36)` and for the _Sony IMX708_ something like `'imx708_wide' (/base/axi/pcie@120000/rp1/i2c@88000/imx708@1a)`.
+
+#### Buffer allocation issues (exception `Failed to allocate buffers`)
+
+If the node fails to allocate buffers, you are probably running out of video memory. This is most likely the case on resource constrained devices with limited memory, such as the Raspberry Pi Zero.
+
+To avoid this, either increase the video memory via `raspi-config`, or decrease the buffer size by reducing the image resolution (parameters `width` and `height`) and/or selecting a chroma subsampled or compressed pixel format that uses less bits per pixel (parameter `format`). For example, a raw format like `XRGB8888` uses 32 bits for each individual pixel, while a chroma subsampled format like `YUYV` only uses 32 bits per two pixels due to the subsampling of the colour components. For details on pixel formats, see the [Linux Kernel Image Formats documentation](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/pixfmt.html).

--- a/README.md
+++ b/README.md
@@ -107,12 +107,12 @@ or dynamically via the [ROS parameter API](https://docs.ros.org/en/rolling/Conce
 
 The camera stream is configured once when the node starts via the following static read-only parameters:
 
-| name              | type                  | description |
-| ----------------- | --------------------- |  ---------- |
-| `camera`          | `integer` or `string` | selects the camera by index (e.g. `0`) or by name (e.g. `/base/soc/i2c0mux/i2c@1/ov5647@36`) [default: `0`]
-| `role`            | `string`              | configures the camera with a `StreamRole` (possible choices: `raw`, `still`, `video`, `viewfinder`) [default: `viewfinder`]
-| `format`          | `string`              | a `PixelFormat` that is supported by the camera [default: auto]
-| `width`, `height` | `integer`             | desired image resolution [default: auto]
+| name              | type                  | description                                                                                                                 |
+| ----------------- | --------------------- | --------------------------------------------------------------------------------------------------------------------------- |
+| `camera`          | `integer` or `string` | selects the camera by index (e.g. `0`) or by name (e.g. `/base/soc/i2c0mux/i2c@1/ov5647@36`) [default: `0`]                 |
+| `role`            | `string`              | configures the camera with a `StreamRole` (possible choices: `raw`, `still`, `video`, `viewfinder`) [default: `viewfinder`] |
+| `format`          | `string`              | a `PixelFormat` that is supported by the camera [default: auto]                                                             |
+| `width`, `height` | `integer`             | desired image resolution [default: auto]                                                                                    |
 
 
 The configuration is done in the following order:

--- a/README.md
+++ b/README.md
@@ -110,7 +110,7 @@ The camera stream is configured once when the node starts via the following stat
 | name              | type                  | description |
 | ----------------- | --------------------- |  ---------- |
 | `camera`          | `integer` or `string` | selects the camera by index (e.g. `0`) or by name (e.g. `/base/soc/i2c0mux/i2c@1/ov5647@36`) [default: `0`]
-| `role`            | `string`              | configures the camera with a `StreamRole` (possible choices: `raw`, `still`, `video`, `viewfinder`) [default: `video`]
+| `role`            | `string`              | configures the camera with a `StreamRole` (possible choices: `raw`, `still`, `video`, `viewfinder`) [default: `viewfinder`]
 | `format`          | `string`              | a `PixelFormat` that is supported by the camera [default: auto]
 | `width`, `height` | `integer`             | desired image resolution [default: auto]
 

--- a/README.md
+++ b/README.md
@@ -71,7 +71,7 @@ ros2 launch camera_ros camera.launch.py
 
 ## Interfaces
 
-The camera node interfaces are compatible with the `image_pipeline` stack. The node publishes the camera images and camera parameters and provides a service to set the camera parameters.
+The camera node interfaces are compatible with the [`image_pipeline`](https://github.com/ros-perception/image_pipeline) stack. The node publishes the camera images and camera parameters and provides a service to set the camera parameters.
 
 ### Topics
 

